File: apps/api/app/schemas/perceptron.py
from typing import Literal

from pydantic import BaseModel

DatasetName = Literal["linearly_separable", "xor"]


class InitializeRequest(BaseModel):
    dataset: DatasetName
    learning_rate: float = 0.1


class PerceptronStateModel(BaseModel):
    weights: list[float]
    bias: float
    learning_rate: float
    epoch: int = 0
    dataset: DatasetName


class TrainRequest(BaseModel):
    state: PerceptronStateModel
    steps: int = 1

File: apps/web/lib/types/perceptron.ts
export type DatasetName = "linearly_separable" | "xor";

export type PerceptronPoint = {
  x1: number;
  x2: number;
  label: number;
};

export type PerceptronPrediction = {
  x1: number;
  x2: number;
  pred: number;
};

export type PerceptronState = {
  weights: [number, number];
  bias: number;
  learning_rate: number;
  epoch: number;
  dataset: DatasetName;
};

export type PerceptronResponse = {
  dataset: {
    points: PerceptronPoint[];
  };
  state: PerceptronState;
  metrics: {
    accuracy: number;
    mistakes: number;
  };
  predictions: PerceptronPrediction[];
  boundary: {
    w1: number;
    w2: number;
    b: number;
  };
  notes: string[];
};
